satellite/gracefulexit/endpoint.go

```diff
@@ -0,0 +1,390 @@
+					}
+
+					bucketID := []byte(storj.JoinPaths(parts[0], parts[1]))
+					limit, privateKey, err := endpoint.orders.CreateGracefulExitPutOrderLimit(ctx, bucketID, newNodes[0].Id, inc.PieceNum, remote.RootPieceId, remote.Redundancy.GetErasureShareSize())
+					if err != nil {
+						return Error.Wrap(err)
+					}
+
+					transferMsg := &pb.SatelliteMessage{
+						Message: &pb.SatelliteMessage_TransferPiece{
+							TransferPiece: &pb.TransferPiece{
+								PieceId:             limit.Limit.PieceId,
+								AddressedOrderLimit: limit,
+								PrivateKey:          privateKey,
+							},
+						},
+					}
+					err = stream.Send(transferMsg)
+					if err != nil {
+						return Error.Wrap(err)
+					}
+					pending.Store(limit.Limit.PieceId, pendingTransfer{
+						path:             inc.Path,
+						pieceSize:        pieceSize,
+						satelliteMessage: transferMsg,
+					})
+				}
+			}
+		}
+		return nil
+	})
+
+	for {
+		// if there are no more transfers and the pending queue is empty, send complete
+		if atomic.LoadInt32(&morePiecesFlag) == 0 && pendingLength() == 0 {
+			// TODO check whether failure threshold is met before sending completed
+			// TODO needs exit signature
+			transferMsg := &pb.SatelliteMessage{
+				Message: &pb.SatelliteMessage_ExitCompleted{
+					ExitCompleted: &pb.ExitCompleted{},
+				},
+			}
+			err = stream.Send(transferMsg)
+			if err != nil {
+				return Error.Wrap(err)
+			}
+			break
+		}
+		// skip if there are none pending
+		if pendingLength() == 0 {
+			continue
+		}
+
+		request, err := stream.Recv()
+		if err != nil {
+			return eofHandler(err)
+		}
+
+		switch m := request.GetMessage().(type) {
+		case *pb.StorageNodeMessage_Succeeded:
+			err = endpoint.handleSucceeded(ctx, &pending, nodeID, m)
+			if err != nil {
+				return Error.Wrap(err)
+			}
+
+		case *pb.StorageNodeMessage_Failed:
+			err = endpoint.handleFailed(ctx, &pending, nodeID, m)
+			if err != nil {
+				return Error.Wrap(err)
+			}
+		default:
+			return Error.New("unknown storage node message: %v", m)
+		}
+	}
+
+	if err := group.Wait(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (endpoint *Endpoint) sendPiecesToTransfer(ctx context.Context, stream processStream, nodeID storj.NodeID) (err error) {
+	defer mon.Task()(&ctx)(&err)
+	endpoint.log.Debug("sending pieces to transfer.", zap.String("nodeID", nodeID.String()))
+
+	return nil
+}
+
+func (endpoint *Endpoint) handleSucceeded(ctx context.Context, pending *sync.Map, nodeID storj.NodeID, message *pb.StorageNodeMessage_Succeeded) (err error) {
+	defer mon.Task()(&ctx)(&err)
+	if message.Succeeded.GetAddressedOrderLimit() == nil {
+		return Error.New("Addressed order limit cannot be nil.")
+	}
+	if message.Succeeded.GetOriginalPieceHash() == nil {
+		return Error.New("Original piece hash cannot be nil.")
+	}
+	endpoint.log.Debug("transfer succeeded.", zap.String("piece ID", message.Succeeded.AddressedOrderLimit.Limit.PieceId.String()))
+
+	// TODO validation
```

Comment:
Check `satellite/metainfo/metainfo.go` -> `filterValidPieces` for a guide to validate the order limits and signatures.